src/modules/cspec/CSPECInstrument.cpp

```diff
@@ -76,74 +101,109 @@ void CSPECInstrument::processReadouts(void) {
       dumpReadoutToFile(readout);
     }
 
+    if(readout.RingId > Conf.MaxRing){
+      XTRACE(DATA, ERR, "Invalid Ring ID: %u, Max Ring ID: %u", readout.RingId, Conf.MaxRing);
+      counters.RingErrors++;
+      continue;
+    }
+
+    if(readout.FENId > Conf.MaxFEN){
+      XTRACE(DATA, ERR, "Invalid FEN ID: %u, Max FEN ID: %u", readout.FENId, Conf.MaxFEN);
+      counters.FENErrors++;
+      continue;
+    }
+
     XTRACE(DATA, DEB, "readout: Phys RingId %d, FENId %d, VMM %d, Channel %d, TimeLow %d",
            readout.RingId, readout.FENId, readout.VMM, readout.Channel, readout.TimeLow);
 
-    //counters.RingRx[readout.RingId]++;
-
     // Convert from physical rings to logical rings
     // uint8_t Ring = readout.RingId/2;
+    uint8_t AsicId = readout.VMM & 0x1;
+    uint8_t HybridId = floor(readout.VMM/2);
```

Comment:
Check bit mask value - check how it's done in freia
(readout.VMM & 0xF) >> 1 instead of floor(readout.VMM/2)